### PR TITLE
IGNITE-17198 Minimal implementation of pure in-memory storage (with in-memory RAFT)

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-17198

### CODE REVIEW

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.raft.server;
+
+/**
+ * Options specific to a Raft group that is being started.
+ */
+public class RaftGroupOptions {
+    private final boolean volatileStores;
```
Comment (SammyVimes) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java:24]: I always forget: we can omit comments for private members, right?

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/VolatileJRaftServiceFactory.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+        return new LocalSnapshotStorage(uri, raftOptions);
+    }
+
+    @Override public RaftMetaStorage createRaftMetaStorage(final String uri, final RaftOptions raftOptions) {
+        return new VolatileRaftMetaStorage();
+    }
+
+    @Override public LogEntryCodecFactory createLogEntryCodecFactory() {
+        return LogEntryV1CodecFactory.getInstance();
```
Comment (SammyVimes) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/VolatileJRaftServiceFactory.java:57]: Probably should be default or something. I don't see that we are ever going to use different codecs (probably just the implementation of this one)

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/VolatileRaftMetaStorage.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+package org.apache.ignite.raft.jraft.storage.impl;
+
+import org.apache.ignite.raft.jraft.entity.PeerId;
+import org.apache.ignite.raft.jraft.option.RaftMetaStorageOptions;
+import org.apache.ignite.raft.jraft.storage.RaftMetaStorage;
+import org.apache.ignite.raft.jraft.storage.VolatileStorage;
+
+/**
+ * Volatile (in-memory) implementation of {@link RaftMetaStorage}. Used for Raft groups storing partition data of
+ * volatile (in-memory) storages.
```
Comment (SammyVimes) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/VolatileRaftMetaStorage.java:27]: I think this `(in-memory)` can be omitted

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 4]
```diff
@@ -500,6 +502,10 @@ private void updateAssignmentInternal(ConfigurationNotificationEvent<byte[]> ass
         CompletableFuture.allOf(futures).join();
     }
 
+    private RaftGroupOptions groupOptionsForInternalTable(InternalTable internalTbl) {
+        return new RaftGroupOptions(internalTbl.storage().isVolatile());
```
Comment (SammyVimes) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:506]: Is it possible that there is going to be more options inside the RaftGroupOptions? Maybe it's not a good idea to use constructor here and you already have factory methods. Maybe you can add some new method like "RaftGroupOptions#optionsForTable(isVolatile)"? Does it make sense?

**Review by SammyVimes — CHANGES_REQUESTED** [review]
Looks good overall, just a couple of minor things

File: .idea/codeStyles/codeStyleConfig.xml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 <component name="ProjectCodeStyleConfiguration">
   <state>
-    <option name="PREFERRED_PROJECT_CODE_STYLE" value="GoogleStyle" />
+    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Ignite 3" />
```
[the remark this answers is not in the document]
Comment (rpuch) [reply .idea/codeStyles/codeStyleConfig.xml]: Oops, good catch!

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLozaTest.java [hunk 6]
```diff
@@ -66,7 +67,7 @@ public class ItLozaTest {
      */
     private RaftGroupService startClient(String groupId, ClusterNode node, Loza loza) throws Exception {
         return loza.prepareRaftGroup(groupId,
-                List.of(node), () -> mock(RaftGroupListener.class)
+                List.of(node), () -> mock(RaftGroupListener.class), RaftGroupOptions.defaults()
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLozaTest.java]: I tried to import it statically in tests (to make it less verbose), but I did not import it statically in production code (to make it more visible). Now, I made all the code to conform to this rule.

WDYT?

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/impl/JraftServerImpl.java [hunk 7]
```diff
@@ -339,13 +351,20 @@ public synchronized boolean startRaftGroup(String groupId, @NotNull RaftGroupEve
...
+            nodeOptions.setRaftMetaUri(serverDataPath.resolve("meta").toString());
+        }
         nodeOptions.setSnapshotUri(serverDataPath.resolve("snapshot").toString());
 
         nodeOptions.setFsm(new DelegatingStateMachine(lsnr));
 
         nodeOptions.setRaftGrpEvtsLsnr(evLsnr);
 
+        if (groupOptions.volatileStores()) {
+            // override to volatile stores
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/server/impl/JraftServerImpl.java]: Sure, removed it

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    public static RaftGroupOptions forPersistentStores() {
+        return new RaftGroupOptions(false);
+    }
+
+    public static RaftGroupOptions forVolatileStores() {
+        return new RaftGroupOptions(true);
+    }
+
+    public RaftGroupOptions(boolean volatileStores) {
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java]: Made it private

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java (hunk 1 above)
Comment (rpuch) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java:25]: In Ignite3 we can, but in this case I preferred to add Javadocs

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/VolatileJRaftServiceFactory.java (hunk 2 above)
Comment (rpuch) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/VolatileJRaftServiceFactory.java]: Moved the methods' implementations to `default` methods

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLozaTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLozaTest.java]: 👍

**Pull request merged**